src/snapshot/mod.rs

```diff
@@ -0,0 +1,242 @@
+        GgrsComponentSnapshotCopyPlugin, GgrsComponentSnapshotReflectPlugin,
+        GgrsEntitySnapshotPlugin, GgrsResourceChecksumHashPlugin, GgrsResourceMapEntitiesPlugin,
+        GgrsResourceSnapshotClonePlugin, GgrsResourceSnapshotCopyPlugin,
+        GgrsResourceSnapshotReflectPlugin, GgrsSnapshotSetPlugin, LoadWorldSet, SaveWorldSet,
+    };
+}
+
+/// Typical [`Resource`] used to store snapshots for a [`Resource`] `R` as the type `As`.
+/// For most types, the default `As = R` will suffice.
+pub type GgrsResourceSnapshots<R, As = R> = GgrsSnapshots<R, Option<As>>;
+
+/// Typical [`Resource`] used to store snapshots for a [`Component`] `C` as the type `As`.
+/// For most types, the default `As = C` will suffice.
+pub type GgrsComponentSnapshots<C, As = C> = GgrsSnapshots<C, GgrsComponentSnapshot<C, As>>;
+
+/// Collection of snapshots for a type `For`, stored as `As`
+#[derive(Resource)]
+pub struct GgrsSnapshots<For, As = For> {
+    /// Queue of snapshots, newest at the front, oldest at the back.
+    snapshots: VecDeque<As>,
+    /// Queue of snapshots, newest at the front, oldest at the back.
+    /// Separate from snapshots to avoid padding.
+    frames: VecDeque<i32>,
+    /// Maximum amount of snapshots to store at any one time
+    depth: usize,
+    _phantom: PhantomData<For>,
+}
+
+impl<For, As> Default for GgrsSnapshots<For, As> {
+    /// Create a default [`GgrsSnapshots`] resource. This will only track a maximum of 8 snapshots.
+    /// If you require a longer rollback window, use [`set_depth`](`GgrsSnapshots::set_depth`)
+    fn default() -> Self {
+        // 8 is the current default `max_prediction_window`
+        let depth = 8;
+
+        Self {
+            snapshots: VecDeque::with_capacity(depth),
+            frames: VecDeque::with_capacity(depth),
+            depth,
+            _phantom: default(),
+        }
+    }
+}
+
+impl<For, As> GgrsSnapshots<For, As> {
+    pub fn set_depth(&mut self, depth: usize) -> &mut Self {
+        self.depth = depth;
+
+        // Greedy allocation to avoid allocating at a more sensitive time.
+        if self.snapshots.capacity() < self.depth {
+            let additional = self.depth - self.snapshots.capacity();
+            self.snapshots.reserve(additional);
+        }
+
+        if self.frames.capacity() < self.depth {
+            let additional = self.depth - self.frames.capacity();
+            self.frames.reserve(additional);
+        }
+
+        self
+    }
+
+    pub const fn depth(&self) -> usize {
+        self.depth
+    }
+
+    pub fn push(&mut self, frame: i32, snapshot: As) -> &mut Self {
+        debug_assert_eq!(
+            self.snapshots.len(),
+            self.frames.len(),
+            "Snapshot and Frame queues must always be in sync"
+        );
+
+        loop {
+            let Some(&current) = self.frames.front() else {
+                break;
+            };
+
+            // TODO: Handle wrapping behavior (wrapping_sub, etc.)
```

Comment:
is this still an open todo? what exactly is meant here?

Reply:
Yes I missed that! I wanted to add some simple logic to handle the possibility of the `i32` frame count wrapping. I know that's a highly unlikely event to actually occur during gameplay, but I still wanted to make sure it was handled.